wasm_memsafety_rtlib.c

```diff
@@ -64,66 +70,183 @@ static void __wasm_memsafety_table_resize(HashTable *table, size_t capacity) {
     table->capacity = capacity;
 }
 
-static bool __wasm_memsafety_table_set(HashTable *table, void *ptr, size_t size) {
+static bool __wasm_memsafety_table_set(HashTable *table, void *ptr, size_t requested_size, size_t allocated_size) {
     if (table->count + 1 > table->capacity * TABLE_MAX_LOAD) {
         size_t capacity = table->capacity < 8 ? 8 : table->capacity * 2;
         __wasm_memsafety_table_resize(table, capacity);
     }
 
     TableEntry *entry = __wasm_memsafety_table_find(table->entries, table->capacity, ptr);
     bool isNewKey = entry->key == NULL;
-    if (isNewKey && entry->size == 0) {
+    if (isNewKey && entry->allocated_size == 0) {
         table->count++;
     }
 
     entry->key = ptr;
-    entry->size = size;
+    entry->requested_size = requested_size;
+    entry->allocated_size = allocated_size;
     return isNewKey;
 }
 
 static TableEntry __wasm_memsafety_table_remove(HashTable *table, void *ptr) {
     if (table->count == 0) {
-        return (TableEntry) {NULL, 0};
+        return (TableEntry) {NULL, 0, 0};
     }
 
     TableEntry *entry = __wasm_memsafety_table_find(table->entries, table->capacity, ptr);
     if (entry->key == NULL) {
-        return (TableEntry) {NULL, 0};
+        return (TableEntry) {NULL, 0, 0};
     }
 
     TableEntry result = *entry;
 
     // Place a tombstone in the entry
     entry->key = NULL;
-    entry->size = 1;
+    entry->requested_size = 0;
+    entry->allocated_size = 1;
 
     return result;
 }
 
 static HashTable table = {.capacity = 0, .count = 0, .entries = NULL};
 
-void *__wasm_memsafety_malloc(size_t align, size_t size) {
-    align = align > 16 ? align : 16;
-    size_t aligned_size = (size + (align - 1)) & (~(align - 1));
+
+// The descriptions for malloc, calloc, realloc and free have been taken from
+// the malloc(3) linux manual page.
+
+#define MTE_ALIGNMENT 16
+#define STRIP_MTE_TAG(ptr) ((void *) ((size_t) (ptr) & 0xF0FFFFFFFFFFFFFF))
+
+static inline size_t align_size(size_t align, size_t size) {
+    // Calculate smallest multiple of align that is not less than size
+    return (size + (align - 1)) & (~(align - 1));
+}
+
+// The malloc() function allocates size bytes and returns a pointer to the
+// allocated memory. The memory is not initialized. If size is 0, then malloc()
+// returns either NULL, or a unique pointer value that can later be
+// successfully passed to free().
+void *__wasm_memsafety_malloc(size_t align, size_t requested_size) {
+    // If we don't early exit here on requested size 0, we would allocate
+    // 16 bytes unnecessarily.
```

Comment:
I don't think this is true. We would still allocate 0 bytes, but I agree that exiting early and returning `NULL` is good.